core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -1560,50 +1567,83 @@ public void run() {
+          CallCredentials callCredentials;
+          if (channelCreds == null) {
+            transportFactory = originalTransportFactory;
+            callCredentials = null;
+          } else {
+            SwapChannelCredentialsResult swapResult =
+                originalTransportFactory.swapChannelCredentials(channelCreds);
+            if (swapResult == null) {
+              this.delegate = Grpc.newChannelBuilder(target, channelCreds);
+              return;
+            } else {
+              transportFactory = swapResult.transportFactory;
+              callCredentials = swapResult.callCredentials;
+            }
+          }
+          ClientTransportFactoryBuilder transportFactoryBuilder =
+              new ClientTransportFactoryBuilder() {
+                @Override
+                public ClientTransportFactory buildClientTransportFactory() {
+                  return transportFactory;
+                }
+              };
+          delegate = new ManagedChannelImplBuilder(
+              target,
+              channelCreds,
+              callCredentials,
+              transportFactoryBuilder,
               new FixedPortProvider(nameResolverArgs.getDefaultPort()));
-          managedChannelImplBuilder.executorPool = executorPool;
-          managedChannelImplBuilder.offloadExecutorPool = offloadExecutorHolder.pool;
         }
 
         @Override
         protected ManagedChannelBuilder<?> delegate() {
-          return managedChannelImplBuilder;
-        }
-
-        @Override
-        public ManagedChannel build() {
-          // TODO(creamsoup) prevent main channel to shutdown if oob channel is not terminated
-          return new ManagedChannelImpl(
-                  managedChannelImplBuilder,
-                  originalTransportFactory,
-                  backoffPolicyProvider,
-                  balancerRpcExecutorPool,
-                  stopwatchSupplier,
-                  Collections.<ClientInterceptor>emptyList(),
-                  timeProvider);
+          return delegate;
         }
       }
 
       checkState(!terminated, "Channel is terminated");
 
       @SuppressWarnings("deprecation")
-      ResolvingOobChannelBuilder builder = new ResolvingOobChannelBuilder(target)
+      ResolvingOobChannelBuilder builder = new ResolvingOobChannelBuilder(target, creds)
           .nameResolverFactory(nameResolverFactory);
 
       return builder
-          .overrideAuthority(getAuthority())
+          .overrideAuthority(ManagedChannelImpl.this.authority())
+          // TODO(zdapeng): executors should not outlive the parent channel.
+          .executor(executor)
+          .offloadExecutor(offloadExecutorHolder.getExecutor())
           .maxTraceEvents(maxTraceEvents)
           .proxyDetector(nameResolverArgs.getProxyDetector())
           .userAgent(userAgent);
     }
 
+    @Override
+    public ChannelCredentials getUnsafeChannelCredentials() {
+      return originalChannelCreds;
```

Comment:
If originalChannelCreds is null, then this should throw.

Reply:
I feel it will be awkward to try and catch a particular type of exception when using the API.
What about allowing `Helper. getChannelCredentials()` and `Helper.getChannelCredentials()` to return `null` if no channel creds were provided by the user?

Reply:
We discussed over chat and agreed to make a new "fake" ChannelCredentials type in ManagedChannelImpl and return it here. When it sees it passed in to create a resolving OOB channel it will use the current transport factory.

It is possible to have Netty and OkHttp always provide valid ChannelCredentials. But something like InProcess in more unclear. It could have a "fake" credential just like we are talking about with ManagedChannelImpl. But for the moment that doesn't buy us much. We can easily do it in the future.

Reply:
As discussed offline. Replaced the return value with a dummy nonnull creds for the moment. The ultimate solution is to let all ChannelBuilder implementations provide a nonnull creds to the ManagedChannelImpl.